### DISCUSSION TITLE
How to contribute to internationalization?

### DISCUSSION THREAD

**Discussion by ghost** [tone-row/flowchart-fun#180] [Q&A] [answered]
Do you use an external service to generate the locales (like [Transifex](https://www.transifex.com)) or do you copy, paste, modify and generate by hand?

**Comment by rob-gordon** [comment, marked as answer]
@RickStanley I typically use https://gengo.com/ to source translations. I use https://github.com/lingui/js-lingui for tokenization and locales. If you want to contribute don't hesitate to take out a PR. You can get a good sense of what's going on in the https://github.com/tone-row/flowchart-fun/tree/main/app/src/locales folder, but if you have any questions just let me know!

**Reply by ghost** [reply]
I started translating the texts to Brazilian Portuguese by copying a locale (e.g. en/messages.po) and manually changing the texts but it felt wrong. 😅

I think it would be nice to have a service to automate this process like [Transifex](https://www.transifex.com/)[^1], I think they provide open source basic features, not sure. Anyway, it's just a suggestion.

Another suggestion is to add translation steps, if needed.

[^1]: Example: [yay localization](https://www.transifex.com/yay-1/yay/) of [yay project](https://github.com/Jguer/yay)

**Reply by rob-gordon** [reply]
Portuguese support would be really cool! Sorry, my previous comment wasn't very clear 🤗 

Here are the steps to add a language:

1. add the language code to app/.linguirc
2. Add language to app/src/locales/i18n.tsx so option appears in the app UI
2. Run `yarn workspace app extract` – this will create the app/src/locales/pt-br/messages.po file
3. Fill out missing strings in app/src/locales/pt-br/messages.po
4. Run `yarn workspace app compile` to generate app/src/locales/pt-br/messages.js
5. Run `yarn start` to make sure it works

I went ahead and [opened a PR](https://github.com/tone-row/flowchart-fun/pull/181) where I did steps 1 to 3. If you're still interested in adding it you can commit directly to the PR (if github will let you 🤔 ), or you may have to fork it. I hope this helps!

**Comment by comradekingu** [comment]
@RickStanley Happy to help if you use Weblate in some capacity.
Gengo reportedly isn't a good place to source translations for software.
